rfcs/007-background-workers.md

```diff
@@ -0,0 +1,287 @@
+to themselves for execution. They do that by setting the `WorkerId` in the jobs
+table entry to their worker ID, removing the job from pending queue and create
+a new entry in their own `"active"` jobs area.
+
+When a job finishes running, either because it was a successful completion or a
+terminal failure, the worker MUST remove the job from its active table and
+clear its `WorkerId` field. Then, based on type-specific behavior, it MUST do
+either one of:
+
+ - Update the `JobState` to `"completed"` and leave the job in the jobs table.
+   It would be up to the job creation logic to inspect and remove it.
+ - Delete the job from the system altogether. This may be useful when running
+   `_replicate` jobs, for example.
+
+
+If a user decided to cancel a job that is running, they MUST toggle `CancelReq`
+to `true`. Then they MUST wait for the job to stop running. If the job is not
+running, they may directly remove the job from the pending and the main jobs
+table.
+
+### Worker Capacity and Long Running Jobs
+
+Some background jobs, like continuous replication jobs, might run for years
+without interruption once they start executing. In that case, workers MAY
+choose to periodically stop some jobs and re-insert them into the pending queue
+for execution and choose to run other jobs which have been waiting. This
+behavior is job type specific.
+
+Worker MAY also choose to use type-specific algorithms to decide how often and
+how many jobs they pull from the pending queue. For example, they could use a
+dampening scheme where the longer their active queue grows the slower the rate
+at which they pull jobs from the queue. That behavior is also job type
+specific.
+
+Another advanced scheme workers could be to start workers on demand based on
+the pending queue backlog. When the backlog subsides, they could be stopped
+(with appropriate dampening factors to reduce flapping). That is again job type
+specific, and is outside the scope of this RFC.
+
+## Worker Lifecyle
+
+Workers can be dynamically started and stopped. When workers start, they MUST
+register themselves in the `"workers"` subspace based on their `WorkerType`.
+Each change to the workers list, should be accompanied by updating the
+`"workers_vs"` versionstamp. That versionstamp MAY be efficiently monitored for
+changes using a watch.
+
+
+### Worker Liveliness Monitoring
+
+As long as the workers are active and ready to process jobs, they MUST
+periodically ping their health key. The period MUST be less than the
+`HealthTimeout` value. Other worker nodes, for example, one of the "neighbors"
+(to left or to the right) of the worker, will monitor that health entry. If a
+worker dies unexpectedly, the neighbor MUST remove all of the dead worker's
+jobs from the active table and submit them back into the pending queue.
+
+
+### Liveliness Race Condition
+
+In case workers come back alive after not updating their health entry often
```

Comment:
Maybe we should reword this to when a worker starts up and during each health update, they must check that their jobs haven't been returned to the queue.

Reply:
I added that workers should check if they are still part of the active workers membership list (they can listen to the workers_vs viewstamp with a watch). If all of the sudden they find themselves removed from the list they should stop running all their jobs, re-initialize and go back to the job queue to pick new jobs.